lib/Monads/trace/Trace_More_RG.thy

```diff
@@ -0,0 +1,685 @@
+  apply (cases xs, simp_all)
+  apply wp
+  done
+
+lemma validI_set_take_helper:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. \<forall>x \<in> set (xs rv s0 s). Q x rv s0 s\<rbrace>
+   \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. \<forall>x \<in> set (take (n rv s0 s) (xs rv s0 s)). Q x rv s0 s\<rbrace>"
+  apply (erule rg_strengthen_post)
+  apply (clarsimp dest!: in_set_takeD)
+  done
+
+lemma whenE_throwError_tsp:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> whenE Q (throwError e) \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. \<not> Q \<and> P s0 s\<rbrace>, -"
+  apply (simp add: whenE_def)
+  apply (intro conjI impI; wp)
+  done
+
+lemma weaker_rg_ifE:
+  assumes x: "\<lbrace>P \<rbrace>,\<lbrace>R\<rbrace> a \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+  assumes y: "\<lbrace>P'\<rbrace>,\<lbrace>R\<rbrace> b \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+  shows      "\<lbrace>P and P'\<rbrace>,\<lbrace>R\<rbrace> if test then a else b \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+  apply (rule rg_weaken_preE)
+   apply (wp x y)
+  apply simp
+  done
+
+lemma twp_split_const_if:
+  assumes x: "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+  assumes y: "\<lbrace>P'\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q'\<rbrace>"
+  shows "\<lbrace>\<lambda>s0 s. (S \<longrightarrow> P s0 s) \<and> (\<not> S \<longrightarrow> P' s0 s)\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. (S \<longrightarrow> Q rv s0 s) \<and> (\<not> S \<longrightarrow> Q' rv s0 s)\<rbrace>"
+  by (cases S, simp_all add: x y)
+
+lemma twp_split_const_ifE_R:
+  assumes x: "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+  assumes y: "\<lbrace>P'\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q'\<rbrace>,\<lbrace>E\<rbrace>"
+  shows "\<lbrace>\<lambda>s0 s. (S \<longrightarrow> P s0 s) \<and> (\<not> S \<longrightarrow> P' s0 s)\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. (S \<longrightarrow> Q rv s0 s) \<and> (\<not> S \<longrightarrow> Q' rv s0 s)\<rbrace>,\<lbrace>E\<rbrace>"
+  by (cases S, simp_all add: x y)
+
+lemma rg_disj_division:
+  "\<lbrakk> P \<or> Q; P \<Longrightarrow> \<lbrace>R\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>S\<rbrace>; Q \<Longrightarrow> \<lbrace>T\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>S\<rbrace> \<rbrakk>
+   \<Longrightarrow> \<lbrace>\<lambda>s0 s. (P \<longrightarrow> R s0 s) \<and> (Q \<longrightarrow> T s0 s)\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>S\<rbrace>"
+  apply safe
+   apply (rule rg_weaken_pre)
+    apply simp
+   apply simp
+  apply (rule rg_weaken_pre)
+   apply simp
+  apply simp
+  done
```

Comment:
could this be a `fastforce`? If not, I feel at least the urge for `safe; rule rg_weaken_pre; simp`.

Reply:
Thanks for noticing this, `by (fastforce intro: rg_weaken_pre)` works.